libs/datasets/combined_datasets.py

```diff
@@ -47,17 +53,82 @@ class RegionLatestNotFound(IndexError):
     pass
 
 
+RegionMaskOrRegions = NewType("RegionMaskOrRegions", Union[RegionMask, Region, Collection[Region]])
+
+
+@final
+@dataclass(frozen=True)
+class DataSourceAndRegionMasks:
+    """Represents a DataSource class and include/exclude region masks.
+
+    Use function `datasource_regions` to create instance of this class.
+
+    Instances of this class can be used in the same places where the DataSource type (not
+    instances of the DataSource) are used.
+
+    Using this class depends on an existing source of all location_ids. Currently it depends on
+    the existing combined data and is used to produce a new combined data. A recursive data
+    dependency isn't great but works and we don't otherwise have a stable source of locations to
+    use for filtering. Fixing this could be part of
+    https://trello.com/c/2Pa4DMyu/836-trim-down-the-timeseriesindexfields
+    """
+
+    data_source_cls: Type[data_source.DataSource]
+    include: Optional[RegionMaskOrRegions] = None
+    exclude: Optional[RegionMaskOrRegions] = None
+
+    @property
+    def EXPECTED_FIELDS(self):
+        """Returns the same property of the wrapped DataSource class."""
+        return self.data_source_cls.EXPECTED_FIELDS
+
+    @property
+    def SOURCE_NAME(self):
+        """Returns the same property of the wrapped DataSource class."""
+        return self.data_source_cls.SOURCE_NAME
+
+    def make_dataset(self) -> MultiRegionDataset:
+        """Returns the dataset of the wrapped DataSource class, with a subset of the regions."""
+        dataset = self.data_source_cls.make_dataset()
+
+        def _get_location_ids(region_mask_or_regions: RegionMaskOrRegions,) -> Collection[str]:
+            if isinstance(region_mask_or_regions, RegionMask):
+                return region_mask_to_location_ids(region_mask_or_regions)
+            elif isinstance(region_mask_or_regions, Region):
+                return [region_mask_or_regions.location_id]
+            else:
+                return [r.location_id for r in region_mask_or_regions]
+
+        if self.include:
+            dataset = dataset.get_locations_subset(_get_location_ids(self.include))
+        if self.exclude:
+            dataset = dataset.remove_locations(_get_location_ids(self.exclude))
+        return dataset
+
+
+def datasource_regions(
```

Comment:
I'm not sure this provides a ton of value over just the class itself?

Reply:
True, not much, but it does make use of this code look more like that in https://www.dropbox.com/scl/fi/wlchtiqvq0cs9aj5lnbfm/Importing-CAN-scraper-data.paper and enforce that callers explicitly list some intention about region.

Reply:
fine by me - easy enough to change if we desire at any point